### PR TITLE
Apply local configuration to install targets

### LINKED ISSUE #7297: local options do not work with install on local packages (was: Run okay with profiling enabled, but installed exe has NO profiling enabled)
**Describe the bug**
Can't install profiling-enabled exe with cabal 3.2.0.0, while direct running works correctly.

**To Reproduce**

```
$ cabal v2-run --enable-profiling --enable-executable-profiling  exe:els -- +RTS --info                                  
Resolving dependencies...
Up to date
 [("GHC RTS", "YES")
 ,("GHC version", "8.8.3")
 ,("RTS way", "rts_thr_p")
 ,("Build platform", "x86_64-apple-darwin")
 ,("Build architecture", "x86_64")
 ,("Build OS", "darwin")
 ,("Build vendor", "apple")
 ,("Host platform", "x86_64-apple-darwin")
 ,("Host architecture", "x86_64")
 ,("Host OS", "darwin")
 ,("Host vendor", "apple")
 ,("Target platform", "x86_64-apple-darwin")
 ,("Target architecture", "x86_64")
 ,("Target OS", "darwin")
 ,("Target vendor", "apple")
 ,("Word size", "64")
 ,("Compiler unregisterised", "NO")
 ,("Tables next to code", "YES")
 ,("Flag -with-rtsopts", "-maxN31 -qg")
 ]
$ cabal v2-install --enable-profiling --overwrite-policy=always --enable-executable-profiling exe:els && els +RTS --info 
Wrote tarball sdist to
 ...
Resolving dependencies...
Up to date
Copying 'els'
 [("GHC RTS", "YES")
 ,("GHC version", "8.8.3")
 ,("RTS way", "rts_thr")
 ,("Build platform", "x86_64-apple-darwin")
 ,("Build architecture", "x86_64")
 ,("Build OS", "darwin")
 ,("Build vendor", "apple")
 ,("Host platform", "x86_64-apple-darwin")
 ,("Host architecture", "x86_64")
 ,("Host OS", "darwin")
 ,("Host vendor", "apple")
 ,("Target platform", "x86_64-apple-darwin")
 ,("Target architecture", "x86_64")
 ,("Target OS", "darwin")
 ,("Target vendor", "apple")
 ,("Word size", "64")
 ,("Compiler unregisterised", "NO")
 ,("Tables next to code", "YES")
 ,("Flag -with-rtsopts", "-maxN31 -qg")
 ]
$ els +RTS -p
els: the flag -p requires the program to be built with -prof
els: 
els: Usage: <prog> <args> [+RTS <rtsopts> | -RTS <args>] ... --RTS <args>
els: ...
$ 
```

**Expected behavior**
The built executable can have `+RTS -p` enabled

**System information**
 - Operating system

macOS Mojave

 - `cabal`, `ghc` versions

```
$ cabal --version
cabal-install version 3.2.0.0
compiled using version 3.2.0.0 of the Cabal library 
$ ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.8.3
```

**Additional context**

### PR DESCRIPTION
**Pull request by alt-romes**
The target of `cabal install` is not considered to be a local package, which means local configuration (e.g. in cabal.project, or flags like --enable-profiling) does not apply to it.

In 76670ebddeb0a914d816f4420c511c1bc1f044eb, we changed the behaviour to applying the local flags to cabal install targets, but it used the literal target string as a package name to which the flags were additionally applied.

However, `cabal install` targets are NOT necessarily package names, so, e.g., if we did `cabal install exe:mycomp`, the local flags would not apply since "exe:mycomp" is not a recognized /package/.

The solution is to parse the target selectors first, and apply the local flags to the package of the resolved targets.

Fixes #7297, #8909, the install part of #7236, #8529, #7832

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Comment by alt-romes** [comment]
Note for tomorrow: I should probably pass --installdir to `cabal install` in the testsuite in order to clear the warnings out of `cabal.out`

**Review by gbaz — APPROVED** [review]

**Comment by gbaz** [comment]
Great sleuthing!

File: cabal-install/src/Distribution/Client/CmdInstall.hs [hunk 1]
```diff
@@ -641,17 +649,29 @@ installAction flags@NixStyleFlags{extraFlags = clientInstallFlags', ..} targetSt
...
+targetPkgNames
+  :: [PackageSpecifier UnresolvedSourcePackage]
+  -- ^ The local packages, to resolve 'TargetAllPackages' selectors
+  -> TargetSelector
+  -> [PackageName]
+targetPkgNames localPkgs = \case
+  TargetPackage _ pkgIds _ -> map pkgName pkgIds
+  TargetPackageNamed name _ -> [name]
+  TargetAllPackages _ -> map pkgSpecifierTarget localPkgs
+  TargetComponent pkgId _ -> [pkgName pkgId]
```
Comment (mpickering) [inline cabal-install/src/Distribution/Client/CmdInstall.hs:673]: Perhaps this is worthy of a comment about how we selected a component but the top-level options will apply to all components of that package.

Comment (alt-romes) [reply cabal-install/src/Distribution/Client/CmdInstall.hs:673]: I will add one

**Comment by alt-romes** [comment]
We should probably backport this

**Comment by mpickering** [comment]
@alt-romes CI has failed after the rebase, the patch needs to be slightly modified it seems.

**Comment by andreabedini** [comment]
@alt-romes It looks like this is causing issues. From the matrix channel (I can reproduce).
```
$ ./cabal install hindent
No cabal.project file or cabal file matching the default glob './*.cabal' was found.
```
I think the issue is that you cannot call `establishProjectBaseContext` outside of a project. Indeed it is only called by the `withProject` subfunction and not by `withProject`. The function `withProjectOrGlobalConfig` just tries `withProject` and if it fails it tries `withoutProject`. Not the best design.

**Comment by alt-romes** [comment]
> @alt-romes CI has failed after the rebase, the patch needs to be slightly modified it seems.

I noticed now that my rebase force-with-lease deleted the test for remote repository package you added @mpickering.
I've readded it in #9707

**Comment by gbaz** [comment]
I'm confused/concerned about this despite reviewing it and missing this element (for which, apologies)

My understanding was this was intended to do the following "The solution is to parse the target selectors first, and apply the local flags to the package of the resolved targets."

However it _also_ appears to switch `cliConfig` to `baseConfig` which means we take not only command line flags but also project file flags into account. I don't believe this was discussed in the ticket, and I don't think it is correct behavior?

Why was this change made and was it necessary to resolve the linked issue, or is it just an additional change that was considered an improvement and sort of "slid in"?

**Comment by alt-romes** [comment]
@gbaz 
> However it also appears to switch cliConfig to baseConfig which means we take not only command line flags but also project file flags into account (I think?). I don't believe this was discussed in the ticket, and I don't think it is correct behavior?

That's not quite right, before my patch:
```haskell
    baseCliConfig =
      commandLineFlagsToProjectConfig
        globalFlags
        flags{configFlags = configFlags'}
        clientInstallFlags'
    cliConfig = addLocalConfigToTargets baseCliConfig targetStrings
```

The change slid in because the `addLocalConfigToTargets` part only needs to get added by the time that I do add them back, and until then we can work with the base configuration as we don't look at the configuration for any package.


Let me clarify what *local* options are (excerpt from a commit message that didn't make it):
```
Currently, there are three kinds of cabal configurations considered when
determining an option of an `ElaboratedConfiguredPackage`:

* Global configuration, in `.cabal/config`

* Local configuration, in
    - Options passed in the cabal invocation, e.g. `cabal build --enable-executable-dynamic`
    - Fields in the top level `cabal.project`, or `cabal.project.local`, e.g. `extra-include-dirs: /opt/homebrew/include`

    Note thus that top-level cabal.project flags and cli flags are
    treated all together at the same level (`local`).

* Per package configuration, as in

    package HsOpenSSL
      extra-include-dirs: /opt/homebrew/Cellar/openssl@3/3.2.0_1/include
      extra-lib-dirs: /opt/homebrew/Cellar/openssl@3/3.2.0_1/lib

Then, we have a definition for whether a package is local to the
project. The local packages are the packages listed in the project which
have a specific source package, rather than just being listed by name in
a `source-repository-stanza`, or in a `package <package-name>` stanza
that configures installed packages.
```

So, *local* options are both top-level cabal.project fields and cli arguments.

Now, in https://github.com/haskell/cabal/commit/76670ebddeb0a914d816f4420c511c1bc1f044eb, the behaviour started to be: apply *local* options to the targets of `cabal install <target>`. However, it used whatever the <target> was as a literal package name, and applied the local options to that package. This is wrong e.g. `exe:myexe` is definitely not a package name, or `all:exes`...
It wasn't particularly my decision to apply local configuration to cabal install targets. I just fixed behaviour which was inconsistent (e.g. `cabal install mypkg --enable-profiling` would already install a profiled executable, where as `cabal install --enable-profiling`, or `cabal install myexe --enable-profiling`, or `cabal install all:exes --enable-profiling` would not)

What did this patch actually break, and that @andreabedini fixed subsequently?
Basically, when we specify `cabal install all:exes`, or `all` or `all:tests`, we need to determine all of the packages local to the project in order to install those. I did this by calling `establishProjectBaseContext` -- which is wrong if done outside of a project. The right thing would be to, if outside of the project, pass an empty list of local packages instead (outside of project, `cabal install all` would install nothing...). Andrea's patch simply couples a refactor with that.

**Comment by tomsmeding** [comment]
I don't know the fine details of what kinds of configuration there are, etc., but just reporting as a user here:
- If I'm in a project (i.e. in a directory with a `cabal.project` file), and I `cabal install` without further arguments, surely that configuration must be fully taken into account.
- If I'm in a project (idem) and I e.g. `cabal install shellcheck`, where `shellcheck` doesn't have anything at all to do with the current project, I **do not** expect the `cabal.project` configuration to be taken into account. Global installation shouldn't take local configuration into account. This would also hold if I had added `--lib`, even with `--package-env`, because that would be local installation _with respect to that particular env file_, not with respect to the "current" project.

If you're intending for this to work differently, that's your purview, but I would then at least like a fat warning when using a global installation command in a context where local configuration is used. :)

**Comment by alt-romes** [comment]
@tomsmeding that sounds about right.

If there is a clear way to distinguish between cli arguments and cabal.project options we could probably look at whether the target refers to a package which is local to the project in order to determine if the cabal.project options should apply. But I'm not sure there is.

> If you're intending for this to work differently, that's your purview, but I would then at least like a fat warning when using a global installation command in a context where local configuration is used. :)

FWIW this decision was made in #8779, I have only patched it to work for the remaining `cabal install <target>` forms.

**Comment by gbaz** [comment]
Alright, thanks for the clarification @alt-romes and sorry for the confusion on my part. I had wrongly thought #8779 only applied cli flags and not the project file, but some testing with cabal-3.10.1.0 has convinced me I was wrong. I think this is not ideal, but we can live with it, even long term, and there's no reason to attach figuring out if we want to change this in the future to this current patch. (so i did miss it in a review -- but not of this pr but that pr)

Thanks again for walking though it so clearly!

**Comment by andreabedini** [comment]
> If there is a clear way to distinguish between cli arguments and cabal.project options we could probably look at whether the target refers to a package which is local to the project in order to determine if the cabal.project options should apply.

I believe we have something like that: `partitionToKnownTargetsAndHackagePackages` in `CmdInstall.hs`
